### PR TITLE
Roles for our pathogen-repo-build GitHub Actions workflow, take 3

### PR DESCRIPTION
**Pull request by tsibley**
A collection of templated repo-specific roles for inside a Nextstrain runtime and one role for the GitHub Actions job itself, i.e. outside the runtime.

The inside-the-runtime roles are given pathogen-specific permissions necessary for the ingest and phylogenetic workflows of a pathogen repo.

The outside-the-runtime role is only necessary/used at the moment for access to AWS Batch.

The roles can only be assumed by specific repos when performed by our centralized pathogen-repo-build workflow. While this doesn't completely prevent off-label use by other GitHub Actions workflows launched from a pathogen repo, it does make it more involved to do so, hopefully to the point of discouragement. The associated GitHub repository configuration is managed by Terraform now as well since the customization of the "sub" claim in GitHub Action's OIDC token is tightly coupled to our AWS role trust policies.

Resolves: <https://github.com/nextstrain/infra/issues/4>
Related-to: <https://github.com/nextstrain/private/issues/96>, <https://github.com/nextstrain/.github/pull/81>
Supersedes: <https://github.com/nextstrain/infra/pull/6>, <https://github.com/nextstrain/infra/pull/7>

## Checklist

- [x] Checks pass

### CODE REVIEW

**Comment by tsibley** [comment]
As none of the changes here should effect production usage—they're all additive or make changes to things not used in production—I've gone ahead and deployed the proposed changes (e.g. run `terraform apply`) in order to much more easily test them with https://github.com/nextstrain/.github/pull/81.

**Review by joverlee521 — APPROVED** [review]
Thank you for iterating on this @tsibley 🙏 

It's too bad that we are blocked by AWS to fully automate the management of per pathogen repo permissions, but this is definitely better than manually managing roles/credentials per repo. 

Separate from this PR, we should have an internal guideline for adding new pathogen repos for automation. Seems like the new docs should live in this repo since this is where we'll have to make changes to opt-in new repos, but also seems hidden in the general `infra` repo...

File: env/production/locals.tf
Comment (tsibley) [reply env/production/locals.tf:20]: Yep, add to that map then `terraform apply`.

We could definitely auto-deploy… but that does require having very broad admin level access to AWS available in GitHub Actions, which makes me a little nervous. I'd say stick with manual deploys for now.

Comment (tsibley) [reply env/production/locals.tf:20]: When we do want to auto-deploy, the best way to gain that admin-level access would be via OIDC again, scoped down tightly to this repo (and ditto for auto-deploying nextstrain.org's Terraform config).

**Comment by tsibley** [comment]
> It's too bad that we are blocked by AWS to fully automate the management of per pathogen repo permissions, but this is definitely better than manually managing roles/credentials per repo.

Yeah. Either AWS or GitHub could make possible what we wanted/needed. AWS by adding more flexible claim → tag mapping. GitHub by adding customized claims.

> Separate from this PR, we should have an internal guideline for adding new pathogen repos for automation. Seems like the new docs should live in this repo since this is where we'll have to make changes to opt-in new repos, but also seems hidden in the general `infra` repo...

Indeed. There's several scopes/levels of doc here to consider. I would think to put some brief doc adjacent to `pathogen-repo-build` (e.g. "Usage of this workflow requires the repository is configured in nextstrain/infra…") and some how-to doc here in this repo (e.g. "Adding a pathogen repo"). Additionally, we'll be having [broader "so you want to automate things, huh?" doc](https://github.com/nextstrain/pathogen-repo-guide/issues/10) in the pathogen-repo-guide, right?, and that could provide some overall contextualization of how the pieces fit together. (We could also put this overall doc in the main docs project or our internal wiki.)

**Comment by genehack** [comment]
> Separate from this PR, we should have an internal guideline for adding new pathogen repos for automation. 

🎉 YES PLEASE! I would be happy to help out here, whether by beta-testing new instructions or helping to draft them or whatever. I expect I'll be wanting to do this for `seasonal-cov` before EOW. 

> Seems like the new docs should live in this repo since this is where we'll have to make changes to opt-in new repos, but also seems hidden in the general `infra` repo...

Maybe the `pathogen-repo-guide` needs to grow a "Nextstrain internal" section to document common patterns for setting up automation in repos?

**Comment by tsibley** [comment]
Added documentation here for supporting a new repo: https://github.com/nextstrain/infra/pull/8/commits/90fec81fcad333f91dd67db5fe8be9dd74f6855d

**Comment by tsibley** [comment]
I'm going to merge this. It's already deployed as noted above and won't be used by our repos until we merge https://github.com/nextstrain/.github/pull/81.

**Pull request merged**
